build-osx.sh

```diff
@@ -1,17 +1,208 @@
 #!/bin/sh
+#
+# build-osx.sh is the master script we use to control the multi-step build processes
+#
+# To understand what it does, run
+#
+# ./build-osx.sh --help
 
-if [ ! -f vst3sdk/LICENSE.txt ]; then
-  echo You have not gotten the submodules required to build Surge. Run the following command to get them.
-  echo git submodule update --init --recursive
-  exit
-fi
 
-premake5 xcode4
-if [ -n "$VST2SDK_DIR" ]; then
-	xcodebuild clean -project surge-vst2.xcodeproj
-	xcodebuild build -configuration Release -project surge-vst2.xcodeproj
+help_message()
+{
+    cat << EOHELP
+
+build-osx.sh is the master script we use to control the multi-step build processes
+You can do a lot with xcode but for things like starting releasing and installing
+locally you will need to interact with this script
+
+To see this message run
+
+   ./build-osx.sh --help
+
+To do a simple local build which is not installed do
+
+   ./build-osx.sh
+
+Options are:
+
+        --help                   Show this message
+        --premake                Run premake only
+
+        --build                  Run the builds without cleans
+        --install-local          Once assets are built, install them locally
+        --build-and-install      Build and install the assets
+        --build-validate-au      Build and install the audio unit then validate it
+
+        --clean                  Clean all the builds
+        --clean-all              Clean all the builds and remove the xcode files and target directories
+
+        --uninstall-surge        Uninstall surge both from user and system areas. Be very careful!
+
+The default behaviour without arguments is to clean and rebuild everything.
+
+Environment variables are:
+
+   VST2SDK_DIR=path      If this points at a valid VST2 SDK, VST2 assets will be built
+   BREWBUILD=TRUE        Uses LLVM clang rather than xcode. If you are XCode < 9.4 you will need this
+
+EOHELP
+}
+
+
+prerequisite_check()
+{
+    if [ ! -f vst3sdk/LICENSE.txt ]; then
+        echo You have not gotten the submodules required to build Surge. Run the following command to get them.
+        echo git submodule update --init --recursive
+        exit 1
+    fi
+
+    # TODO: Check premake is installed
+}
+
+run_premake()
+{
+    premake5 xcode4
+    touch Surge.xcworkspace/premake-stamp
+}
+
+run_premake_if()
+{
+    if [[ premake5.lua -nt Surge.xcworkspace/premake-stamp ]]; then
+        run_premake
+    fi
+}
+
+run_clean()
+{
+    flavor=$1
+    xcodebuild clean -configuration Release -project surge-${flavor}.xcodeproj
+}
+
+run_build()
+{
+    flavor=$1
+    xcodebuild build -configuration Release -project surge-${flavor}.xcodeproj
+}
+
+default_action()
+{
+    run_premake
+    if [ -n "$VST2SDK_DIR" ]; then
```

Comment:
Since we're in a green field, I think would be prudent to check for the project directory
```
if [ -d surge-vst2.xcodeproj ]; then
```
everywhere instead of looking for the `env` variable

Reply:
Done!